frontend/public/components/utils/container-select.tsx

```diff
@@ -0,0 +1,92 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import {
+  Divider,
+  MenuToggle,
+  MenuToggleElement,
+  Select,
+  SelectGroup,
+  SelectList,
+  SelectOption,
+} from '@patternfly/react-core';
+import { useTranslation } from 'react-i18next';
+
+import { ContainerModel } from '@console/internal/models';
+import { ContainerSpec } from '@console/internal/module/k8s';
+import { ResourceName } from './resource-icon';
+
+export const getContainerLabel = (name: string) => (
+  <ResourceName name={name} kind={ContainerModel.kind} />
+);
+
+const getSelectOptions = (containers: Containers) =>
+  Object.values(containers).map(({ name }) => (
+    <SelectOption key={name} value={name}>
+      {getContainerLabel(name)}
+    </SelectOption>
+  ));
+
+export const ContainerSelect: React.FC<ContainerSelectProps> = ({
+  containers,
+  currentKey = '',
+  initContainers = {},
+  onChange,
+}) => {
+  const [isOpen, setIsOpen] = React.useState(false);
+  const [selected, setSelected] = React.useState<string>(
+    currentKey || Object.values(containers)[0].name,
+  );
+  const { t } = useTranslation();
+
+  if (_.isEmpty(containers)) {
+    return null;
+  }
+
+  const onToggleClick = () => {
+    setIsOpen(!isOpen);
+  };
+  const onSelect = (_event: React.MouseEvent<Element, MouseEvent> | undefined, value: string) => {
+    onChange(value);
+    setSelected(value);
+    setIsOpen(false);
+  };
+  const containerOptions = getSelectOptions(containers);
+
+  return (
+    <Select
+      isOpen={isOpen}
+      selected={selected}
+      onSelect={onSelect}
+      onOpenChange={(open) => setIsOpen(open)}
+      toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
+        <MenuToggle ref={toggleRef} onClick={onToggleClick} isExpanded={isOpen}>
+          {getContainerLabel(selected)}
+        </MenuToggle>
+      )}
+      shouldFocusToggleOnSelect
+    >
+      {!_.isEmpty(initContainers) ? (
+        <>
+          <SelectGroup label={t('public~Containers')}>
+            <SelectList>{containerOptions}</SelectList>
+          </SelectGroup>
+          <Divider />
+          <SelectGroup label={t('public~Init containers')}>
+            <SelectList>{getSelectOptions(initContainers)}</SelectList>
+          </SelectGroup>
+        </>
+      ) : (
+        <SelectList>{containerOptions}</SelectList>
+      )}
+    </Select>
+  );
+};
+
+type Containers = { [key: string]: ContainerSpec };
+
+type ContainerSelectProps = {
+  containers: Containers;
```

Comment:
This could be follow-on tech-debt, but `containers` and `initContainers` should be arrays. There is a lot of unnecessary logic involved in transforming these props back and forth between arrays and objects.

Reply:
I noticed the same thing, but opted to keep the goofy objects since the existing API was using them, and I didn't want to change the API since the goal here was the swap the underlying components.